### ISSUE TITLE
AE2 Networks and Copy-Paste gadget Kills the TPS.

### ISSUE THREAD

**Issue by ShadoWZYE** [Direwolf20-MC/BuildingGadgets#422]
Run into a little server breaking issue this morning, it seems like my Copy- Paste tool got bound to my ME when i was attempting to fast build a few CPU's and didn't work. Anyhow, today i needed the tool to patch some holes and My TPS went straight bellow 2 as soon as i got it in hand <did all the testing>. I can reproduce the same issue as many times as i want on the world i'm playing on. Binding the tool to a chest resolves the problem. Did some poking into creative as well, it seems that it might have something to do with the number of different items available into the bind inventory <The ME in this case> Tested with creative dirt/gravel and flint available into the ME and everything was stable.

Building Gadgets 2.7.4 on Enigmatica 2 Expert V1.72b

I can provide a copy of my world for testing if needed.

TPS shown bellow <i had to take the gadget off of hand to get the command to run>
![2019-12-10_16 29 19](https://user-images.githubusercontent.com/48412812/70551954-6c91c100-1b70-11ea-967b-c39639d4b9a9.png) 
latest.log: 
[latest (1).log](https://github.com/Direwolf20-MC/BuildingGadgets/files/3946267/latest.1.log)

**Comment by MichaelHillcox** [comment]
Yeah, we've had another issue like this. Could you by any chance send me the world and coords? I don't have a way to test and debug this without a world to figure out what's going wrong.

**Comment by ShadoWZYE** [comment]
For sure <expected as much, replicating an ME with a few millions of items ain't a 5 minutes work>, i'll have to upload it to my google drive though coz' it's pretty big for github, you'll find my compact machine base at x:-64 y:228 z:201 overworld. Entering that CM will drop you right into my ME Interface room. (access to the entire AE2 network). I should mention that there's no Quantum links or anything alike. I have 1 backbone network running my p2p's. No subnets. So, what's visible on the terminals is pretty much all the stuff that exists. There's one auxiliar network i used just for the purpose of creating singularities <ae2> since import/export buses are somewhat faster than other item transport methods, that network is standalone.

Possibly some memory leak. That or the gadget is trying to do something funny with all the items in that ME. Since i haven't had a look at the code for that binding and what's it doing can't really say anything more.

PS: The ME being in a different dimension is not the cause of the issue, tested that already :) Also sorry for the late response.

[https://drive.google.com/open?id=10ZpwRPWHa_jLJ4Lr4N__NAsM42vOnLiQ]

**Comment by MajorTuvok** [comment]
Notice that the Gadget is iterating once over all the Items to index them, however as this index is not cached this is done once per render...

This iteration is performed so that whilst building or rendering, checking an Item is expected constant time - one should be able to speed that up, by implementing an AE handle and an AE Provider, which could make use of the AE System already knowing it's types and then directly adding them...

**Comment by ShadoWZYE** [comment]
That constant polling done on the AE might be the culprit of the issue.
I've uploaded my world save to GitHub. (We'll linked it better said)please
feel free to have a look around. But honestly I'm almost sure it has to do
with the amount of different items in the network and the indexing
frequency. As I said, it's quite a big network with lot's of things going
on (it might not look like it due to the way it's set up).

**Comment by MajorTuvok** [comment]
I deleted this comment with a purpose: that only applies to 1.14

**Comment by MajorTuvok** [comment]
In particular it iterates the whole remote inventory per missing item check (here)[https://github.com/Direwolf20-MC/BuildingGadgets/blob/763f1adc6ff35857638b84a7420b2d901d746799/src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java#L109].

**Comment by MichaelHillcox** [comment]
I'm 100% sure ae2 has done all the hard work for this. We just need to make sure we're using their methods for their inventory and not just generic iteration. We've know about this for a while but without an ae2 system full on items, we've not been able to performance test. Thanks @ShadoWZYE for the download. I'll see what I can figure out. I'm pretty busy atm and basically the only Dev on the project at this time so Dev has come to a crawl for a while... Back to it soon :D

**Comment by ShadoWZYE** [comment]
@MichaelHillcox  my pleasure. Good luck with it. :)

**Comment by ShadoWZYE** [comment]
`              for (int i = 0; i < remoteInventory.getSlots(); i++) {
                ItemStack containerItem = remoteInventory.getStackInSlot(i);
                if (containerItem.getItem() == itemStack.getItem() && containerItem.getMetadata() == itemStack.getMetadata() && containerItem.getCount() >= count) {
                    remoteInventory.extractItem(i, count, false);
                    return true;
                }
            }
        }`
This , code wise on AE is not funny. Maybe it would be possible to detect the type of inventory and index according to that, for ae and other storage mods alike <ex: refined storage, rftools etc..> this is going to cause issues at a certain point. All dependent on the items in those networks. Using the mods built in item handler for interrogation regarding a specific item type and count would be the best way to deal with it, caching the response <?> though that might cause issues on systems with moving items, so the cache can be skipped. 
In other words, searching every tick trough all the items in a massive "chest" to find dirt is not elegant nor efficient at all.

**Comment by MichaelHillcox** [comment]
Yeah, this isn't a good system that wasn't though about fully on its original implementation. I think a whitelist of mods and known containers might be more ideal whilst I work out the specifics

**Comment by ShadoWZYE** [comment]
The functionality is really great, being able to print huge buildings
straight from the AE is neat. Whilst you can move that is :))) Minor
undesired results with huge storages, nothing that can be fixed. Looking
forward for the upcoming versions :)

**Comment by MichaelHillcox** [comment]
I've tracked the source of this bug done but I don't have time to implement the fix at the moment unfortunatly.

**Comment by MichaelHillcox** [comment]
We thank you for your issue report, unfortunatly we no longer support this version of Minecraft / Our mod. I've added your issue to a collection of issues that the community is welcome to create pull requests for. Hopefully your issue is resolved in time 👍 Master list: #535
